components/camel-spring/src/main/java/org/apache/camel/spring/CamelContextFactoryBean.java

```diff
@@ -341,43 +338,49 @@ protected void initPropertyPlaceholder() throws Exception {
         }
     }
 
-    public void onApplicationEvent(ApplicationEvent event) {
-        // From Spring 3.0.1, The BeanFactory applicationEventListener 
-        // and Bean's applicationEventListener will be called,
-        // So we just delegate the onApplicationEvent call here.
+    @Override
+    public void start() {
+        try {
+            setupRoutes();
+        } catch (Exception e) {
+            throw wrapRuntimeCamelException(e);
+        }
+
+        // when the routes are setup we need to start the Camel context
+        context.start();
+    }
 
-        SpringCamelContext context = getContext(false);
+    @Override
+    public void stop() {
         if (context != null) {
-            // we need to defer setting up routes until Spring has done all its dependency injection
-            // which is only guaranteed to be done when it emits the ContextRefreshedEvent event.
-            if (event instanceof ContextRefreshedEvent) {
-                try {
-                    setupRoutes();
-                } catch (Exception e) {
-                    throw wrapRuntimeCamelException(e);
-                }
-            }
-            // let the spring camel context handle the events
-            context.onApplicationEvent(event);
-        } else {
-            LOG.debug("Publishing spring-event: {}", event);
-
-            if (event instanceof ContextRefreshedEvent) {
-                // now lets start the CamelContext so that all its possible
-                // dependencies are initialized
-                try {
-                    // we need to defer setting up routes until Spring has done all its dependency injection
-                    // which is only guaranteed to be done when it emits the ContextRefreshedEvent event.
-                    setupRoutes();
-                    LOG.trace("Starting the context now");
-                    getContext().start();
-                } catch (Exception e) {
-                    throw wrapRuntimeCamelException(e);
-                }
-            }
+            context.stop();
         }
     }
 
+    @Override
+    public void stop(Runnable callback) {
+        stop();
```

Comment:
Yeah it will end with a [warning in the log file](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/support/DefaultLifecycleProcessor.java#L259), and I guess that the [default logic in Spring](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/support/DefaultLifecycleProcessor.java#L374-L378) depends on not calling the runnable in case of exceptions. The [javadoc](http://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/SmartLifecycle.html#stop-java.lang.Runnable-) unfortunately does not mention an error scenario.